Neos.Flow/Classes/Mvc/Controller/AbstractController.php

```diff
@@ -253,33 +268,66 @@ protected function forwardToRequest(ActionRequest $request)
      * @param array $arguments Array of arguments for the target action
-     * @param integer $delay (optional) The delay in seconds. Default is no delay.
-     * @param integer $statusCode (optional) The HTTP status code for the redirect. Default is "303 See Other"
-     * @param string $format The format to use for the redirect URI
+     * @param int $delay (optional) The delay in seconds. Default is no delay.
+     * @param int $statusCode (optional) The HTTP status code for the redirect. Default is "303 See Other"
+     * @param string|null $format The format to use for the redirect URI
      * @return void
-     * @throws StopActionException
+     * @throws StopActionException | UnsupportedRequestTypeException | NoMatchingRouteException
      * @see forward()
      * @api
      */
     protected function redirect($actionName, $controllerName = null, $packageKey = null, array $arguments = [], $delay = 0, $statusCode = 303, $format = null)
     {
-        if ($packageKey !== null && strpos($packageKey, '\\') !== false) {
-            list($packageKey, $subpackageKey) = explode('\\', $packageKey, 2);
+        // The namespaced request arguments are not yet handled by the new ActionUriSpecification (see below), so we are
+        // going to use the legacy implementation for any case, where we would need namespaces, which is always true,
+        // if the current request is a "sub request"
+        // Note: You need to compare request with request->getMainRequest, because action requests are NEVER main
+        // requests
+        if ($this->request !== $this->request->getMainRequest()) {
+            if ($packageKey !== null && strpos($packageKey, '\\') !== false) {
+                list($packageKey, $subpackageKey) = explode('\\', $packageKey, 2);
+            } else {
+                $subpackageKey = null;
+            }
+            $this->uriBuilder->reset();
+            if ($format === null) {
+                $this->uriBuilder->setFormat($this->request->getFormat());
+            } else {
+                $this->uriBuilder->setFormat($format);
+            }
+
+            $uri = $this->uriBuilder->setCreateAbsoluteUri(true)->uriFor($actionName, $arguments, $controllerName, $packageKey, $subpackageKey);
+            $this->redirectToUri($uri, $delay, $statusCode);
+        }
+
+        if ($packageKey === null) {
+            $packageKey = $this->request->getControllerPackageKey();
+            $subpackageKey = $this->request->getControllerSubpackageKey();
+        } elseif (str_contains($packageKey, '\\')) {
+            [$packageKey, $subpackageKey] = explode('\\', $packageKey, 2);
         } else {
             $subpackageKey = null;
         }
-        $this->uriBuilder->reset();
-        if ($format === null) {
-            $this->uriBuilder->setFormat($this->request->getFormat());
-        } else {
-            $this->uriBuilder->setFormat($format);
+
+        $targetActionUriSpecification = ActionUriSpecification::create(
```

Comment:
or  @fcool  and me discussed `NameSpacedActionUriSpecification` for subrequests

Reply:
but i rather not just duplicate all the Uribuilder::mergeArgumentsWithRequestArguments into another place like: NameSpacedActionUriSpecification

we wouldnt gain anything.

what could be cool, if we can swap out all the internals of the uribuilder to use the ActionUriSpecification or NameSpacedActionUriSpecification so we internally use the new api. But i would wait for this refactor untill 9.0 as it could be breaky in edgecases

since our rewrite method in the abstract controller didnt really got better, i would be fine with continuing to use the old uribuilder here for now and wait for 9.0 to introduce the  NameSpacedActionUriSpecification.